### ISSUE TITLE
dim() on 0-column data.table produced in non-data.table-aware package is wrong

### ISSUE THREAD

**Issue by akersting** [Rdatatable/data.table#2422]
If a data.table, which is passed to a function of a non-data.table-aware package, is subsetted there, such that a 0-column data.table/data.frame is produced, `dim()` on that data.table/data.frame falsely reports 0 rows.

```
library(data.table)
X <- data.table(a = 1:10)

# imitate subsetting in function of non-data.table-aware package
Y <- `[.data.frame`(X, , character(), drop = FALSE)

dim.data.frame(Y)  # returns c(10, 0)
dim(Y)  # returns c(0, 0), also in function of non-data.table-aware package
```

```
> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 8 (jessie)

Matrix products: default
BLAS: /usr/lib/atlas-base/atlas/libblas.so.3.0
LAPACK: /usr/lib/atlas-base/atlas/liblapack.so.3.0

locale:
 [1] LC_CTYPE=de_DE.UTF-8       LC_NUMERIC=C               LC_TIME=de_DE.UTF-8        LC_COLLATE=de_DE.UTF-8    
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=de_DE.UTF-8    LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.10.4-2

loaded via a namespace (and not attached):
[1] tools_3.4.2 yaml_2.1.14
```

**Comment by MichaelChirico** [comment]
Actually, `Y` _does_ have dimension `(0, 0)` (i.e., the error is not `dim`'s fault):

```
dput(Y)
# structure(list(), .Names = character(0), 
#           class = c("data.table", "data.frame"), row.names = c(NA, -10L))
```

In fact, `Y` is not internally consistent, since `row.names` has retained the 10-row structure, but the table itself is empty. 

```
print(Y)
# Null data.table (0 rows and 0 cols)
```

**Comment by mb706** [comment]
The bug is not whether or not the "actual" dimension is `(0, 0)`, but whether or not it *should* be. `data.table` represents zero-row-nonzero-column tables just fine (as a nonempty list of zero length vectors), but does not go on to representing nonzero-row-zero-column tables as empty lists of nonempty vectors (think about it). In that sense, `Y`s internal state is entirely consistent, `data.table` arguably just chose to interpret it as dimension (0, 0).

The behaviour of `data.frame` in these cases, which is
```r
> iris[character(0)]
data frame with 0 columns and 150 rows
```
has very desirable properties, for example one has
```
cbind(df[x], df[y]) === df[c(x, y)]
```
Even though `X by 0` and `0 by X` `data.frame`s or matrices contain no data, they make edge case behaviour more consistent and are useful for package development (even though they may not help much in interactive sessions).

**Comment by jangorecki** [comment]
I understand your point, but still I prefer `c(0, 0)` as correct answer. Rows are _childs_ of columns, if there are no columns there should be no rows returned. As Michael pointed out, it looks more like bug in R. This is a little bit problematic because there is not much control over how non-data.table-aware package will process user data. Eventually good solution would be to handle this edge case by detecting if call like `df[character()]` (resulting in 0 cols data.table) was made from non-data.table package and then make an exception.

**Comment by mb706** [comment]
> Rows are childs of columns, if there are no columns there should be no rows returned

That is just a detail about implementation, though. `data.table` does handle zero row but nonzero column tables just fine (`dt[integer(0), ]`). An R `matrix` can also have zero rows *xor* zero columns (although in the underlying representation, as a simple vector with additional info of dimensionality, the data is a 0-length vector).

Having zero column tables with nonzero rows is similar to having zero-length vectors that still have a type.

I changed the confusing `===` notation.

**Comment by jangorecki** [comment]
> Having zero column tables with nonzero rows is similar to having zero-length vectors that still have a type.

My understanding is

Having zero **rows** table with nonzero **columns** is similar to having zero-length vectors that still have a type.

Also my comment from linked issue:
> In columnar storage row is a child of column. Without column no rows exists. This make sense for multidimensional structures like vector/matrix/array but not data.frames.

**Comment by heavywatal** [comment]
In other words,
Row subsetting should keep column number.
Column subsetting should keep row number.

``` r
tbl = tibble::as_tibble(iris)
dt = data.table::data.table(iris)

ncol(tbl[seq_len(3L),])
#> [1] 5
ncol(tbl[integer(0L),])
#> [1] 5
ncol(dt[seq_len(3L),])
#> [1] 5
ncol(dt[integer(0L),])
#> [1] 5

nrow(tbl[,"Species"])
#> [1] 150
nrow(tbl[,FALSE])
#> [1] 150
nrow(dt[,"Species"])
#> [1] 150
nrow(dt[,FALSE])        # Surprise!
#> [1] 0
```

<sup>Created on 2019-02-06 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by heavywatal** [comment]
OK, let's forget about `row.names` (I don't like it either) and focus on `dim(x)[1L]`. I am currently working on a thin igraph wrapper with Rcpp. Edges and vertices themselves are stored in `igraph_t` object, and their attributes such as names and weights are stored in data.frames, say, `Eattr` and `Vattr`. Their row numbers should always remain the same as the edge and vertex numbers, respectively. In this senario, it is quite natural to start from (and sometimes shrink to) zero-column nonzero-row data.frames. If it was not allowed, I would have to switch two different methods for adding a new column to a non-empty data.frame and for adding a first column to a dim `c(0, 0)` data.frame or null placeholder.

**Comment by akersting** [comment]
> data.frame is not a multidimensional data structure of any particular dimension

I strongly disagree. A data.frame (and hence also a data.table) is a two-dimensional data structure . That is also why `dim` returns a vector with two elements.

Anyway, there are two issues with the current behavior of data.table:
1. It breaks non-data.table-aware packages.
2. It is inconsistent in itself, since subsetting on one dimension (columns) might change the other one (rows).

**Comment by jangorecki** [comment]
@akersting 

> > data.frame is not a multidimensional data structure of any particular dimension

> I strongly disagree. A data.frame (and hence also a data.table) is a two-dimensional data structure .

data.frame is two-dimensional data structure but not a (any particular) case of multidimensional data. Because of that we can store different data types in different columns. This is not possible for multidimensional data where column is no different from row, page or any other name you will use instead of integer sequence that maps data into dimensions. Names like `rows`, `columns`, `pages` doesn't really have meaning for multidimensional data, they only maps an integer dimension indexes in some visual representation. This is also the reason why applying transpose function for multidimensional data will never alter the data but only re-arrange, which is not true for data.frames where transpose can alter data.

**Comment by jangorecki** [comment]
related discussion: https://stat.ethz.ch/pipermail/r-devel/2019-May/077796.html

**Comment by nbenn** [comment]
I just ran into this issue and while I understand (and agree there is some merit to) @jangorecki's "nestedness" argument, I still feel the current data.table behavior is counter-intuitive. Out of the big three data frame structures in R (the other two being `tibble` and base R `data.frame`), `data.table` is the only one to interpret a zero-col/nonzero-row data frame in this way. I feel it would make for a smoother user-experience if zero-col/nonzero-row `data.table` were to become possible.

**Comment by fkohrt** [comment]
I think I fail to see instances where the current behaviour can be considered a feature, not a bug. I created a wrapper class around data.tables that pre-allocates rows (until #660 gets resolved) and not having zero-column data.tables with non-zero rows further complicates code: I have to delay the creation of the underlying data.table until users have provided at least one column. And all pre-allocated rows vanish once the last column in the data.table is removed. I really wish it would be different, but I started at least documenting the current behavior via #5615.

**Comment by jan-glx** [comment]
If you want to continue thinking of a `data.table`s as a `list` of `vectors` instead of something matrix like (which could have `dim(3,0)`), shouldn't `dim` return `c(NA, 0)` for an empty zero column `data.table`? 

To be a bit more constructive:

I often summarize data like this:

```r
iris <- as.data.table(iris)
iris[, .N, by="Species")][, .N:=NULL][] # to get a data.table with the unique Species values
```
 I know, I could could also do this:
```r
iris[, .unique(.SD), .SDcols="Species"]
```
but if later realize that I actually need some summary its harder to get back
```r
iris[, .(n_obs = .N), by="Species")]
```
ideally I'd like to be able to do something
```r
iris[, .(), by="Species")]
```
but I need to tell data.table about the `nrow()==1`
```r
as.data.table(iris)[, data.frame(row.names = 1), by=.(Species)]
```
does not work. If `data.table` would support non-zero-row-zero-column `data.tables`  there would be a natural way to express this.
